### PR TITLE
Fixed the display of actionable proposals page when a single Sns fails

### PR DESCRIPTION
**Pull request by mstrasinskis**
# Motivation

Currently, when loading actionable proposals for a single Sns fails, the actionable proposals page displays a loading state. This happens because the loading store only compares successfully loaded Snses and ignores failures. This PR addresses the issue by modifying the `actionableProposalsLoadedStore` to account for failed responses.

Note: the proposals page for a failed SNS remains unchanged and displays the loading state indefinitely.

# Changes

- New `failedActionableSnsesStore`.
- Use it in `actionableSnsProposalsByUniverseStore` to calculate the loading state.

# Tests

- `failedActionableSnsesStore`
- `actionableProposalsLoadedStore` uses `failedActionableSnsesStore`.
- `loadActionableSnsProposals` updates `failedActionableSnsesStore`.

Tested locally by simulating the failing call with adding `throw new Error()` inside the [queryProposals](https://github.com/dfinity/nns-dapp/blob/02ceec03e01235b6926004a619eed1a342882bb9/frontend/src/lib/api/sns-governance.api.ts#L550).

# Todos

- [x] Add entry to changelog (if necessary).

# Screenshots

## Before

image

## After

image

## Failed Sns proposals page (w/o changes)

image

### CODE REVIEW

File: frontend/src/lib/services/actionable-sns-proposals.services.ts [hunk 1]
```diff
@@ -80,6 +83,8 @@ export const loadActionableProposalsForSns = async (
     });
   } catch (err) {
     console.error(err);
+
+    get(failedActionableSnsesStore).push(rootCanisterId.toText());
```
Comment (dskloetd) [inline frontend/src/lib/services/actionable-sns-proposals.services.ts:87]: This doesn't work. It won't update subscribers. You need to call `set` or `update` on the store.

Comment (mstrasinskis) [reply frontend/src/lib/services/actionable-sns-proposals.services.ts]: Good catch, udpated.

File: frontend/src/lib/services/actionable-sns-proposals.services.ts [hunk 2]
```diff
@@ -80,6 +83,9 @@ export const loadActionableProposalsForSns = async (
     });
   } catch (err) {
     console.error(err);
+
+    // Store the failed root canister ID to provide the correct loading state.
+    failedActionableSnsesStore.add(rootCanisterId.toText());
```
Comment (dskloetd) [inline frontend/src/lib/services/actionable-sns-proposals.services.ts:88]: Is it a problem that we only add to the store and never remove? Should we remove in the success path? Or is it not possible to call `loadActionableProposalsForSns` more than once?

File: frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts [hunk 3]
```diff
@@ -398,6 +401,32 @@ describe("actionable proposals derived stores", () => {
 
       expect(get(actionableProposalsLoadedStore)).toEqual(true);
     });
+
+    it("should respect failed snses as well", async () => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts:405]: What does "respect" mean? Can the description be more explicit?
Is it `"should consider failed snses as loaded"`?

File: frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts [hunk 4]
```diff
@@ -398,6 +401,32 @@ describe("actionable proposals derived stores", () => {
...
+      actionableSnsProposalsStore.set({
+        rootCanisterId: principal0,
+        proposals: [],
+        includeBallotsByCaller: false,
+      });
+      actionableNnsProposalsStore.setProposals([mockProposalInfo]);
+
+      expect(get(actionableProposalsLoadedStore)).toEqual(false);
+
+      failedActionableSnsesStore.add(principal1.toText());
```
Comment (dskloetd) [inline frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts:426]: Please make sure to reset `failedActionableSnsesStore` in `beforeEach`.

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts [hunk 5]
```diff
@@ -115,8 +120,25 @@ describe("actionable-sns-proposals.services", () => {
         proposals,
         include_ballots_by_caller: [true],
       }) as SnsListProposalsResponse;
+    const spyQuerySnsProposalsImplementation = async ({ rootCanisterId }) =>
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts:123]: Why did you move the implementation?
By moving and changing the code at the same time it's really hard to see what changed.

Would it make sense to just be explicit about what we want to return in the individual test. When the logic here gets complicated it's hard to predict what the test will do.

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts [hunk 6]
```diff
@@ -177,26 +189,64 @@ describe("actionable-sns-proposals.services", () => {
...
+      });
+      expect(spyQuerySnsProposals).toHaveBeenCalledWith({
+        identity: mockIdentity,
+        rootCanisterId: rootCanisterId2,
+        certified: false,
+        params: expectedQuerySnsProposalsFilterParams,
+      });
+
+      // expect a single error to be logged
+      expect(spyConsoleError).toHaveBeenCalledTimes(1);
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts:244]: What error do we expect?

File: frontend/src/tests/lib/stores/actionable-sns-proposals.store.spec.ts [hunk 7]
```diff
@@ -72,3 +75,22 @@ describe("actionableSnsProposalsStore", () => {
...
+
+describe("failedActionableSnsesStore", () => {
+  beforeEach(() => {
+    failedActionableSnsesStore.resetForTesting();
+  });
+
+  it("should store root canister ids", () => {
+    failedActionableSnsesStore.add("1");
+    failedActionableSnsesStore.add("2");
+    expect(get(failedActionableSnsesStore)).toEqual(["1", "2"]);
```
Comment (dskloetd) [inline frontend/src/tests/lib/stores/actionable-sns-proposals.store.spec.ts:87]: Can you expect it to be empty before anything is added?

File: frontend/src/lib/services/actionable-sns-proposals.services.ts (hunk 2 above)
Comment (mstrasinskis) [reply frontend/src/lib/services/actionable-sns-proposals.services.ts:89]: In theory, it’s possible that the actionables failed (due to network conditions), but later, after navigating to the normal proposals, the user will vote and reload the actionables. It’s a somewhat artificial scenario, but to ensure we haven’t missed anything, let’s be precise and update the failed store after a successful attempt.

File: frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts (hunk 3 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts]: Thanks, updated.

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts (hunk 5 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts]: Sorry and thanks for the explanation. I undo the refactoring.

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts (hunk 6 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts:246]: Added.

File: frontend/src/tests/lib/stores/actionable-sns-proposals.store.spec.ts (hunk 7 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/stores/actionable-sns-proposals.store.spec.ts:88]: Done.

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts [hunk 8]
```diff
@@ -244,6 +244,10 @@ describe("actionable-sns-proposals.services", () => {
 
       // expect a single error to be logged
       expect(spyConsoleError).toHaveBeenCalledTimes(1);
+      expect(spyConsoleError).toBeCalledWith(
+        new Error("sns query proposals test fail")
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts:248]: Add
```
const snsQueryError = new Error("sns query proposals test fail");
```
on line 206
and use it here (line 248) and on line 214.

File: frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts (hunk 4 above)
Comment (dskloetd) [reply frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts:427]: Looks done.

**Review by dskloetd — APPROVED** [review]
Thanks!

File: frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts (hunk 8 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/actionable-sns-proposals.services.spec.ts]: Good point, looks nicer.

File: frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts (hunk 4 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/derived/actionable-proposals.derived.spec.ts:427]: Sorry, missed this comment, but yes, it's done.
